File: MATLAB/src/utility/singleMeshSuperquadrics.m
function [x_mesh, y_mesh, z_mesh] = singleMeshSuperquadrics(x, varargin)

R = eul2rotm(x(6 : 8));
t = x(9 : 11);

% with tapering or not
taper = false;
% color = 'r';
arclength = 0.02;

for k = 1 : size(varargin, 2)
    if strcmp(varargin{k}, 'Taper')
        taper = varargin{k + 1};
    end
%     if strcmp(varargin{k}, 'Color')
%         color = varargin{k + 1};
%     end   
    if strcmp(varargin{k}, 'Arclength')
        arclength= varargin{k + 1};
    end    
end

% validate dimensionality
if taper == true
    if size(x, 2) ~= 13
        error('Input parameters should have dimension (:, 13) for taperred SQ.')
    end
else
    if size(x, 2) ~= 11
        error('Input parameters should have dimension (:, 11) for taperred SQ.')
    end
end

% avoiding numerical instability of points sampling on superquadrics
if x(1) < 0.01 %0.007
    x(1) = 0.01;
end
if x(2) < 0.01
    x(2) = 0.01;
end

[point_eta] = uniformSampledSuperellipse(x(1), [1, x(5)], arclength);
[point_omega] = uniformSampledSuperellipse(x(2), [x(3), x(4)], arclength);

x_mesh = ones(size(point_omega, 2), size(point_eta, 2));
y_mesh = ones(size(point_omega, 2), size(point_eta, 2));
z_mesh = ones(size(point_omega, 2), size(point_eta, 2));

for m = 1 : size(point_omega, 2)
    for n = 1 : size(point_eta, 2)
        point_temp = [point_omega(:, m) * point_eta(1, n); point_eta(2, n)];
        
        if taper == true
            fx = x(12) * point_temp(3) / x(5) + 1;
            fy = x(13) * point_temp(3) / x(5) + 1;
            fz = 1;
            
            point_temp(1) = point_temp(1) * fx;
            point_temp(2) = point_temp(2) * fy;
            point_temp(3) = point_temp(3) * fz;
        end
        
        point_temp = R * point_temp + t';
        
        x_mesh(m, n) = point_temp(1);
        y_mesh(m, n) = point_temp(2);
        z_mesh(m, n) = point_temp(3);
    end
end

% ---------------------------------utility functions ----------------------
    function [point, theta] = uniformSampledSuperellipse(epsilon, scale, arclength)
        threshold = 1e-2;
        num_limit = 10000;
        theta = zeros(1, num_limit);
        theta(1) = 0;
        
        for i = 2 : num_limit
            dt = dtheta(theta(i - 1), arclength, threshold, scale, epsilon);
            theta_temp = theta(i - 1) + dt;
            
            if theta_temp > pi/4
                break
            else
                if i < num_limit
                    theta(i) = theta_temp;
                else
                    error(['The number of the sampled points exceeds the limit of ', ...
                        num2str(num_limit * 4),...
                        '. Please increase the arclength or raise the limit'])
                end
            end
        end
        critical = i;
        
        for j = critical + 1 : num_limit
            dt = dtheta(theta(j - 1), arclength, threshold, flip(scale), epsilon);
            theta_temp = theta(j - 1) + dt;
            
            if theta_temp > pi/4
                break
            else
                if j < num_limit
                    theta(j) = theta_temp;
                else
                    error(['The number of the sampled points exceeds the limit of ', ...
                        num2str(num_limit * 4),...
                        '. Please increase the arclength or raise the limit'])
                end
            end
        end
        
        num_pt = j - 1;
        theta = theta(1 : num_pt);
        
        points_fw = angle2points(theta(1 : critical - 1), scale, epsilon);
        points_bw = flip(angle2points(theta(critical : end), flip(scale), epsilon), 2);
        point = [points_fw, [points_bw(2, :); points_bw(1, :)]];
        
        point = [point, flip([-point(1, 1 : num_pt - 1); point(2, 1 : num_pt - 1)], 2), ...
            [-point(1, 2 : end); -point(2, 2 : end)], flip([point(1, 1 : num_pt - 1); ...
            -point(2, 1 : num_pt - 1)], 2)];
        
    end

    function [dt] = dtheta(theta, arclength, threshold, scale, sigma)
        if theta < threshold
            dt = abs((arclength / scale(2) + (theta)^(sigma))^(1 / sigma) ...
                - (theta));
        else
            dt = arclength / sigma * ((cos(theta) ^ 2 * sin(theta) ^ 2) / ...
                (scale(1) ^ 2 * cos(theta) ^ (2 * sigma) * sin(theta) ^ 4 + ...
                scale(2) ^ 2 * sin(theta) ^ (2 * sigma) * cos(theta) ^ 4))^(1 / 2);
        end
    end

    function [point] = angle2points(theta, scale, sigma)
        point = zeros(2, size(theta, 2));
        point(1, :) = scale(1) .* sign(cos(theta)) .* abs(cos(theta)).^sigma;
        point(2, :) = scale(2) .* sign(sin(theta)) .* abs(sin(theta)).^sigma;
    end

end

File: src/marching_primitives/single_mesh_superquadrics.py
import numpy as np
from .sdf_superquadric import eul2rotm


def _dtheta(theta, arclength, threshold, scale, sigma):
    """Compute angular step for uniform arclength sampling."""
    if theta < threshold:
        dt = abs((arclength / scale[1] + theta ** sigma) ** (1.0 / sigma) - theta)
    else:
        ct = np.cos(theta)
        st = np.sin(theta)
        dt = arclength / sigma * np.sqrt(
            (ct ** 2 * st ** 2) /
            (scale[0] ** 2 * ct ** (2 * sigma) * st ** 4 +
             scale[1] ** 2 * st ** (2 * sigma) * ct ** 4)
        )
    return dt


def _angle2points(theta, scale, sigma):
    """Convert angles to superellipse points."""
    ct = np.cos(theta)
    st = np.sin(theta)
    points = np.zeros((2, len(theta)))
    points[0] = scale[0] * np.sign(ct) * np.abs(ct) ** sigma
    points[1] = scale[1] * np.sign(st) * np.abs(st) ** sigma
    return points


def _uniform_sampled_superellipse(epsilon, scale, arclength):
    """
    Uniformly sample a superellipse by arclength.

    Parameters
    ----------
    epsilon : float
        Shape parameter.
    scale : array-like, shape (2,)
        Scale factors [a, b].
    arclength : float
        Target arclength between samples.

    Returns
    -------
    point : ndarray (2, M)
        Sampled superellipse points.
    """
    threshold = 1e-2
    num_limit = 10000
    theta = np.zeros(num_limit)

    # Forward pass
    critical = num_limit
    for i in range(1, num_limit):
        dt = _dtheta(theta[i - 1], arclength, threshold, scale, epsilon)
        theta_temp = theta[i - 1] + dt
        if theta_temp > np.pi / 4:
            critical = i
            break
        if i < num_limit - 1:
            theta[i] = theta_temp
        else:
            raise RuntimeError(
                f"Sampled points exceed limit of {num_limit * 4}. "
                "Increase arclength or raise the limit."
            )

    # Backward pass with flipped scale
    scale_flip = scale[::-1]
    num_pt = critical
    for j in range(critical, num_limit):
        dt = _dtheta(theta[j - 1], arclength, threshold, scale_flip, epsilon)
        theta_temp = theta[j - 1] + dt
        if theta_temp > np.pi / 4:
            num_pt = j
            break
        if j < num_limit - 1:
            theta[j] = theta_temp
        else:
            raise RuntimeError(
                f"Sampled points exceed limit of {num_limit * 4}. "
                "Increase arclength or raise the limit."
            )

    theta = theta[:num_pt]

    # Forward points
    points_fw = _angle2points(theta[:critical], scale, epsilon)
    # Backward points (with flipped scale, then flip back)
    points_bw_raw = _angle2points(theta[critical - 1:], scale_flip, epsilon)
    points_bw = points_bw_raw[:, ::-1]  # flip columns
    points_bw = np.array([points_bw[1], points_bw[0]])  # swap rows

    point = np.hstack([points_fw, points_bw])

    # Build full superellipse (4 quadrants)
    n = num_pt - 1
    q2 = np.array([-point[0, :n], point[1, :n]])[:, ::-1]
    q3 = np.array([-point[0, 1:], -point[1, 1:]])
    q4 = np.array([point[0, :n], -point[1, :n]])[:, ::-1]

    point = np.hstack([point, q2, q3, q4])
    return point


def single_mesh_superquadrics(x, arclength=0.02, taper=False):
    """
    Generate a 3D mesh for a single superquadric.

    Parameters
    ----------
    x : array-like, shape (11,) or (13,)
        Superquadric parameters: [eps1, eps2, ax, ay, az, eul_z, eul_y, eul_x, tx, ty, tz]
        With taper: additional [taper_x, taper_y].
    arclength : float
        Arclength sampling resolution.
    taper : bool
        Whether tapering is enabled.

    Returns
    -------
    x_mesh, y_mesh, z_mesh : ndarray
        2D arrays of mesh coordinates.
    """
    x = np.array(x, dtype=float)

    if taper:
        if len(x) != 13:
            raise ValueError("Input parameters should have dimension 13 for tapered SQ.")
    else:
        if len(x) != 11:
            raise ValueError("Input parameters should have dimension 11 for SQ.")

    # Clamp small shape parameters for numerical stability
    if x[0] < 0.01:
        x[0] = 0.01
    if x[1] < 0.01:
        x[1] = 0.01

    R = eul2rotm(x[5:8])
    t = x[8:11]

    point_eta = _uniform_sampled_superellipse(x[0], np.array([1.0, x[4]]), arclength)
    point_omega = _uniform_sampled_superellipse(x[1], np.array([x[2], x[3]]), arclength)

    n_omega = point_omega.shape[1]
    n_eta = point_eta.shape[1]

    x_mesh = np.zeros((n_omega, n_eta))
    y_mesh = np.zeros((n_omega, n_eta))
    z_mesh = np.zeros((n_omega, n_eta))

    for m in range(n_omega):
        for n in range(n_eta):
            pt = np.array([
                point_omega[0, m] * point_eta[0, n],
                point_omega[1, m] * point_eta[0, n],
                point_eta[1, n]
            ])

            if taper:
                fx = x[11] * pt[2] / x[4] + 1.0
                fy = x[12] * pt[2] / x[4] + 1.0
                pt[0] *= fx
                pt[1] *= fy

            pt = R @ pt + t
            x_mesh[m, n] = pt[0]
            y_mesh[m, n] = pt[1]
            z_mesh[m, n] = pt[2]

    return x_mesh, y_mesh, z_mesh
